ipie/addons/adafqmc/utils/miscellaneous.py

```diff
@@ -0,0 +1,325 @@
+        nucnp = comm.bcast(nucnp, root=0)
+        obsconst = comm.bcast(obsconst, root=0)
+        nelec0 = comm.bcast(nelec0, root=0)
+        hcore_shared = get_shared_array(comm, hcoreshape, dtype)
+        if comm.Get_rank() == 0:
+            hcore_shared[:] = hcorenp[:]
+        comm.Barrier()
+        chol_shared = get_shared_array(comm, cholshape, dtype)
+        if comm.Get_rank() == 0:
+            chol_shared[:] = cholnp[:]
+        comm.Barrier()
+        packedchol_shared = get_shared_array(comm, packedcholshape, dtype)
+        if comm.Get_rank() == 0:
+            packedchol_shared[:] = packcholnp[:]
+        comm.Barrier()
+        obsmat_shared = get_shared_array(comm, hcoreshape, dtype)
+        if comm.Get_rank() == 0:
+            obsmat_shared[:] = obsmat[:]
+        comm.Barrier()
+        # Convert shared memory numpy array into pytorch tensors
+        hcore = torch.from_numpy(hcore_shared)
+        chol = torch.from_numpy(chol_shared)
+        packedchol = torch.from_numpy(packedchol_shared)
+        nuc = torch.from_numpy(nucnp)
+        obs = (torch.from_numpy(obsmat_shared), torch.from_numpy(obsconst))
+        hamobs = HamObs(nelec0, hcore.shape[0], hcore, chol, nuc, obs, packedchol, obs_type)
+    else:
+        raise RuntimeError("Shared memory not available")
+    return hamobs
+
+
+def get_hf_wgradient(initialguess, nelec0, ovlp_mat, hamobs, trial_type, coupling):
+    """
+    mo_coeff: molecular orbital coefficients from hartree fock
+    Returns
+    -------
+    trial_wf: trial wavefunction in ao basis
+    """
+    if trial_type == "RHF":
+        _, trial_obs = hartree_fock(
+            initialguess, nelec0, hamobs.h1e + coupling * hamobs.obs[0], ovlp_mat, hamobs.chol
+        )
+        return trial_obs
+    else:
+        raise NotImplementedError
+
+
+# pylint: disable=arguments-differ
+class TrialwithTangent(torch.autograd.Function):
+    """
+    A custom autograd function that stores the trial wavefunction with precomputed trial and tangent
+    """
+
+    @staticmethod
+    def forward(ctx, coupling, trial, tangent):
+        ctx.save_for_backward(tangent)  # Save precomputed gradient for backward pass
+        return trial
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (tangent,) = ctx.saved_tensors  # Retrieve precomputed gradient
+        return tangent * grad_output, None, None
+
+
+class TrialwithTangent_matcoupling(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, coupling, trial, tangent):
+        ctx.save_for_backward(tangent)  # Save precomputed gradient for backward pass
+        return trial
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (tangent,) = ctx.saved_tensors  # Retrieve precomputed gradient
+        return (
+            torch.einsum("abij, ab -> ij", tangent, grad_output),
+            None,
+            None,
+        )  # None for precomputed_grad's grad
+
+
+def trial_tangent(coupling, trial, tangent):
```

Comment:
Why do these functions exist? Why not just use apply?

Reply:
Ah it's sort of like a wrapper

Reply:
It's a little dubious to add a function to call the method itself.

Reply:
It is a trick I used to store part of the gradients so I pass in this function in the ADAFQMC class, so a wrapper will make the code cleaner I thought

Reply:
Ohhhh, this makes sense, sorry. Please add a typehint (Callable) to your tangent arg in AFAFQMC, I missed it expected a callable.